### PR TITLE
Adding System.Threading.Channels for simplified, better performing code.

### PR DESCRIPTION
**Pull request by stebet**
## Proposed Changes

This adds System.Threading.Channels for a much more optimized and easy to use construct for asynchronous consumer/producer queue (like used in the Async Dispatcher and the Recovery queue). After looking at optimizations and code surrounding the AsyncConcurrentQueue this seems like a much more sane approach.

This adds a new dependency on the System.Threading.Channels NuGet package (supported by all platforms).

## Types of Changes

- [ ] Bug fix (non-breaking change which fixes issue #NNNN)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)

## Checklist

- [X] I have read the `CONTRIBUTING.md` document
- [X] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [X] All tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

## Further Comments

This package has been used in previous PRs (f.ex. https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/824) and also the PRs that were part of the bigger Pipelines and allocation PRs, but this stands very well on it's own and is perfectly compatible with both 7.0.0 and 6.X. and is a simple change.

### CODE REVIEW

File: projects/RabbitMQ.Client/client/events/AsyncEventingBasicConsumer.cs [hunk 1]
```diff
@@ -45,12 +45,14 @@ public override async Task HandleBasicConsumeOk(string consumerTag)
...
 
         ///<summary>Fires the Received event.</summary>
-        public override async Task HandleBasicDeliver(string consumerTag, ulong deliveryTag, bool redelivered, string exchange, string routingKey, IBasicProperties properties, ReadOnlyMemory<byte> body)
+        public override Task HandleBasicDeliver(string consumerTag, ulong deliveryTag, bool redelivered, string exchange, string routingKey, IBasicProperties properties, ReadOnlyMemory<byte> body)
         {
-            await base.HandleBasicDeliver(consumerTag, deliveryTag, redelivered, exchange, routingKey, properties, body).ConfigureAwait(false);
-            await (Received?.Invoke(
-                this,
-                new BasicDeliverEventArgs(consumerTag, deliveryTag, redelivered, exchange, routingKey, properties, body)) ?? Task.CompletedTask).ConfigureAwait(false);
+            if (Received != null)
```
Comment (lukebakken) [inline projects/RabbitMQ.Client/client/events/AsyncEventingBasicConsumer.cs:50]: This change raises red flags for me and does not seem to be related to the rest of the changes in this PR. All of the other methods in `AsyncEventingBasicConsumer` are `async`, and use `await` to call the base class methods (which, strangely enough, don't do anything).

**Review by lukebakken — CHANGES_REQUESTED** [review]

File: projects/RabbitMQ.Client/client/events/AsyncEventingBasicConsumer.cs (hunk 1 above)
Comment (stebet) [reply projects/RabbitMQ.Client/client/events/AsyncEventingBasicConsumer.cs:50]: The reason for this change is that the base.HandleBasicDeliver is a no-op, so it's removed. That leaves the method with just one 'await' call and no other operation that takes place. Therefore there is no need to create async state machine and we can just return the task directly and save ourselves some CPU cycles.

However I can see that this might seem strange, and out-of-line with the rest of the code. It's a micro-optimization, which at this point is probably not necessary. I'll change it back.

Comment (lukebakken) [reply projects/RabbitMQ.Client/client/events/AsyncEventingBasicConsumer.cs]: I will make a note to review this code for 7.0, thanks.

**Review by lukebakken — APPROVED** [review]

**Comment by lukebakken** [comment]
Thanks!
